### PR TITLE
Enable span sampling based on `thread.name` and `thread.id`

### PR DESCRIPTION
**Pull request by heyams**
Fix https://github.com/microsoft/ApplicationInsights-Java/issues/3145

Propose `sampling_relevant`: https://github.com/open-telemetry/semantic-conventions/pull/213

In order to enable span sampling based on `thread.name` and `thread.id`, these two thread attributes need to be added before `Span.startSpan()`. 

This proposal doesn't cover the case of manual instrumentation.

Alternatively, a `ThreadDetailsAttributeExtractor` can be used to add thread attributes. That will require changes to all instrumentations to `addAttributesExtractor` when initializing the `Instrumenter`.

### CODE REVIEW

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/Instrumenter.java [hunk 1]
```diff
@@ -43,6 +47,12 @@
  */
 public class Instrumenter<REQUEST, RESPONSE> {
 
+  /** Current &quot;managed&quot; thread ID (as opposed to OS thread ID). */
+  private static final AttributeKey<Long> THREAD_ID = longKey("thread.id");
+
+  /** Current thread name. */
+  private static final AttributeKey<String> THREAD_NAME = stringKey("thread.name");
+
```
Comment (trask) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/Instrumenter.java:55]: can we use `SemanticAttributes.*` constants instead?

File: instrumentation-api/build.gradle.kts [hunk 2]
```diff
@@ -14,6 +14,7 @@ group = "io.opentelemetry.instrumentation"
 dependencies {
   api("io.opentelemetry:opentelemetry-api")
   implementation("io.opentelemetry:opentelemetry-extension-incubator")
+  implementation("io.opentelemetry:opentelemetry-semconv")
```
Comment (trask) [inline instrumentation-api/build.gradle.kts:17]: oh, sorry, I see the problem with my suggestion, we don't want to rely on (unstable) semconv artifact from `instrumentation-api`

go ahead and revert this latest change, and let's wait for @mateuszrzeszutek. I think probably we'll want to introduce a ThreadDetailsAttributesExtractor.

Comment (mateuszrzeszutek) [reply instrumentation-api/build.gradle.kts:17]: > I think probably we'll want to introduce a ThreadDetailsAttributesExtractor.

Yeah, that would be my preferred way to do this. (Also see #4641)

We could just manually add the thread attributes extractor everywhere at first, but I hope that at some point we'd have something like #6651 but for `AttributesExtrator`s -- that's a future thing though.

Comment (heyams) [reply instrumentation-api/build.gradle.kts:17]: yup, i didn't want to add a dependency initially. wasn't thinking about unstable though. i can revert

Comment (heyams) [reply instrumentation-api/build.gradle.kts]: Will create ThreadDetailsAttributesExtractor.

**Comment by heyams** [comment]
https://github.com/open-telemetry/semantic-conventions/pull/213#issuecomment-1708711237

**Pull request closed without merging**